### PR TITLE
Longer tests, logging, warnings

### LINKED ISSUE #51: add longer tests to the test_consistent file
Longer tests will help make sure any cases are covered by consistency checks. 

Use a few different model configurations and random seeds in different checks to increase redundancy in the consistency checking.

[Original suggestion here](https://github.com/DeltaRCM/pyDeltaRCM_WMT/pull/36#pullrequestreview-423541767)

### PR DESCRIPTION
**Pull request by amoodie**
This PR implements some additional consistency testing. I was reluctant to make many tests, as was initially suggested in discussion and #51, because if there are many tests, and you make a known breaking change, it's quite a bit of homework to go back and make all those tests passing again. I added two more tests, such that on a breaking change, the dev would have to go back and change 4 lines to match the new bed elevations. 

I also did work on the i/o file logger, and warnings changes and testing. Previously the logger wrote to a file in the cwd, but now the log is written to the location defined by `out_dir` in the `.yml` config file. Now, every time something happens (just a few markers through the code) the action is written to the log, regardless of the verbosity (only one item is written to log per timestep, so this doesn't appreciably slow things down). The `verbose` flag controls whether these messages are also printed to the console. These actions could be wrapped into a function to simplify the code (I'll open an issue if you all like these changes).

This PR ~will probably close~ relates to but does not complete #51, and addresses most of #45. Between this PR and #52, most of #35 would be implemented too (but still needs to be completely addressed once we rewerite to use hdf in #13).

Will rebase the code once #52  is merged.

### CODE REVIEW

**Comment by ericbarefoot** [comment]
> the dev would have to go back and change 4 lines to match the new bed elevations.

Can this be automated somehow? Like can there be a private function to go and change consistency tests to match the new reality if we make a breaking change we want to pass?

**Comment by amoodie** [comment]
> Can this be automated somehow?

Yeah, that's an interesting idea. It could probably be automated somehow. Automating would require editing the python code via a script though, so would probably need to be a bash script. Something with `grep` to pull the new values from the failed test outputs and then `sed -i ` to replace the new values into the script. But this does not sound like an easy task to me.

Automating that would certainly make it easier to write a ton more tests though.

**Comment by amoodie** [comment]
hi @elbeejay, is this [what you had in mind here](https://github.com/DeltaRCM/pyDeltaRCM_WMT/pull/36#pullrequestreview-423541767)? Or were you thinking like a long, serious, model run? 

The latter could be configured as another "job" on travis, to call some python script that runs the model for a few hundred timesteps (jobs time out at some point...). 

If you mean the latter, then this should be disconnected from #51 before merging, or it will accidentally close the issue. 

Either way, I think this can be merged now if you are good with it :+1:

File: tests/test_shared_tools.py
Comment (elbeejay) [inline tests/test_shared_tools.py:239]: Thanks for flagging this test as well as leaving the segfault issues in the below commented out tests. Will be helpful when debugging to see if these get fixed.

File: tests/test_consistent.py
Comment (elbeejay) [inline tests/test_consistent.py:69]: This is of course longer than a single step test. But I guess my concern is about reproduction of one of the aspects that makes the model unique - the ability to form a distributary channel network. So while this is definitely a step ahead and an improvement to our previous checks, I think we could do better.

**Review by elbeejay — APPROVED** [review]
> hi @elbeejay, is this what you had in mind here? Or were you thinking like a long, serious, model run?
> 
> The latter could be configured as another "job" on travis, to call some python script that runs the model for a few hundred timesteps (jobs time out at some point...).

I like the effort and think this is definitely a step forward for the unit testing. I just can't wrap my head around how we would go about testing the ability of the model (as we continue to alter it) to transition from the initial jet to a reasonable channel network. This (at least to me) is one of the critical things DeltaRCM has to do and there is no obvious or intuitive way to test that without a model run of many timesteps (the point about travis timing out is well taken).

I don't know how we get around it, but that is my concern because if we break the model and it no longer forms a channel network then we have to revert to an older version of the code with no real knowledge of where the 'breaking' change was made.



> If you mean the latter, then this should be disconnected from #51 before merging, or it will accidentally close the issue.

If you don't mind leaving #51 open I think we could improve upon it. Maybe my initializing tests from various pre-developed topographies will help get around running super long tests, not sure just an early thought. 



> Either way, I think this can be merged now if you are good with it :+1:

Yes I agree, these are nice changes that continue to move us forward in the right direction :smile:

File: tests/test_consistent.py
Comment (amoodie) [reply tests/test_consistent.py:69]: yeah, I agree. I'm going to unlink the issue so it stays open.

**Pull request merged**
